### PR TITLE
remote schemas - support for basic schema stitching

### PR DESCRIPTION
**Pull request by ecthiender**
### Description
Now one can add other graphql endpoints as custom resolvers with Hasura, and Hasura will stitch the schema of multiple remote servers with Hasura's schema.

#### Constraint
There cannot be conflicting type names and top-level node names. It will result in remote schema conflict errors.
E.g - type with name `User` both in hasura and remote schema. Same for top-level node names.

#### Limitations in first version
- Cannot mix nodes from different graphql servers in one query/mutation. All top-level nodes has to be from one graphql server
- Subscriptions to remote graphql server are not supported
- Interfaces and unions are not supported, if the remote schema has interfaces there will be errors.
- Types with the exact same name and structure are also conflicts. In the future version, these types will be merged.

These limitations will be worked on in the next version

#### Add remote schema API
Endpoint: `/v1/query`
query:
```json
{
  "type": "add_remote_schema",
  "args": {
    "name": "myremoteschema",
    "comment": "some comment",
    "definition": {
      "url": "http://localhost:3000",
      "url_from_env": "MY_REMOTE_GQL_SERVER",
      "headers": [],
      "forward_client_headers": true
    }
  }
}  
```
`name` is mandatory
`comment` is optional
`url` or `url_from_env` should be present.
`headers` is same as event triggers headers 
`forward_client_headers` is `boolean` field. If set it will forward all headers from the client to the remote server

#### Delete remote schema API
Endpoint: `/v1/query`
query:
```json
{ "type" : "remove_remote_schema",
  "args": {
    "name": "myremoteschema"
   }
}
```

What component does this PR affect? 

- [x] Server
- [ ] Console
- [ ] CLI
- [ ] Docs
- [ ] Community Content
- [ ] Build System

Requires changes from other components? If yes, please mark the components:

- [ ] Server
- [x] Console
- [ ] CLI
- [x] Docs
- [x] Community Content
- [ ] Build System

### Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Docs update
- [ ] Community content

### Checklist:

- [x] I have read the **[contributing guide](https://github.com/hasura/graphql-engine/blob/master/CONTRIBUTING.md)** and my code conforms to the guidelines.
- [x] This change requires a change in the documentation. 
- [ ] I have updated the documentation accordingly.
- [x] I have added required tests.

### CODE REVIEW

File: console/cypress/helpers/remoteSchemaHelpers.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
+export const baseUrl = Cypress.config('baseUrl');
+export const getRemoteSchemaName = (i, schemaName) =>
+  `test-remote-schema-${schemaName}-${i}`;
+export const getRemoteGraphQLURL = () =>
+  'https://hasura-realtime-poll.herokuapp.com/v1alpha1/graphql';
```
Comment (praveenweb) [inline console/cypress/helpers/remoteSchemaHelpers.js]: Update this.

File: console/src/components/Services/CustomResolver/Common/Common.js [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+import {
+  inputChange,
+  UPDATE_FORWARD_CLIENT_HEADERS,
+} from '../Add/addResolverReducer';
+
+import CommonHeader from '../../Layout/ReusableHeader/Header';
+
+const graphqlurl = (
+  <Tooltip id="tooltip-cascade">
+    Remote GraphQL server’s URL. E.g. https://my-domai/v1alpha1/graphql
```
[the remark this answers is not in the document]
Comment (dsandip) [reply console/src/components/Services/CustomResolver/Common/Common.js]: or just https://my-domain.com/graphql ?

File: console/src/components/Services/CustomResolver/CustomResolverRouter.js [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+      <Route path="manage" component={rightBar(connect)}>
+        <IndexRedirect to="schemas" />
+        <Route path="schemas" component={landingCustomResolverGen(connect)} />
+        <Route
+          path="add"
+          component={addConnector(connect)}
+          onEnter={composeOnEnterHooks([migrationRedirects])}
+        />
+        <Route
+          path=":resolverName/details"
```
Comment (praveenweb) [inline console/src/components/Services/CustomResolver/CustomResolverRouter.js]: resolverName to be either changed to resolverId or url encoding should be done.

File: server/src-lib/Hasura/GraphQL/RemoteServer.hs [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+import           Hasura.RQL.Types
+
+import qualified Hasura.GraphQL.Schema         as GS
+import qualified Hasura.GraphQL.Validate.Types as VT
+
+
+introspectionQuery :: BL.ByteString
+introspectionQuery = $(embedStringFile "src-rsr/introspection.json")
+
+fetchRemoteSchema
```
Comment (0x777) [inline server/src-lib/Hasura/GraphQL/RemoteServer.hs]: adding logging

File: server/src-lib/Hasura/RQL/Types/RemoteSchema.hs [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+data RemoteSchemaDef
+  = RemoteSchemaDef
+  { _rsName             :: !Text
+  , _rsUrl              :: !(Either N.URI UrlFromEnv)
+  , _rsHeaders          :: ![HeaderConf]
+  , _rsFwdClientHeaders :: !Bool
+  } deriving (Show, Eq, Lift)
+
+type RemoteSchemaMap = Map.HashMap N.URI RemoteSchemaDef
```
Comment (0x777) [inline server/src-lib/Hasura/RQL/Types/RemoteSchema.hs]: Should be `Map SchemaName RemoteSchemaDef`

File: server/src-lib/Hasura/RQL/Types/RemoteSchema.hs [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+                 ]
+      Right urlFromEnv ->
+        J.object [ "url_from_env" J..= urlFromEnv
+                 , "headers" J..= headers
+                 , "name" J..= name
+                 , "forward_client_headers" J..= fwdHdrs
+                 ]
+
+
+data AddRemoteSchemaQuery
```
Comment (0x777) [inline server/src-lib/Hasura/RQL/Types/RemoteSchema.hs]: can just have three keys `name`, `definition`, `comment`

File: server/src-lib/Hasura/RQL/Types/RemoteSchema.hs (hunk 5 above)
Comment (0x777) [reply server/src-lib/Hasura/RQL/Types/RemoteSchema.hs]: Have a `RemoteSchemaInfo` type which captures the final URI (either directly specified or read from an env var)

File: server/src-lib/Hasura/GraphQL/Schema.hs [hunk 7]
```diff
@@ -1476,14 +1530,85 @@ mkGCtxMapTable tableCache (TableInfo tn _ fields rolePerms constraints pkeyCols
...
+      unless (null conflictedNodes) $
+        throw400 RemoteSchemaConflicts $ nodesMsg conflictedNodes
+    _ -> return ()
+
+  where
+    tyMsg ty = "types: [" <> nodesTxt ty <> "] already exist in current graphql schema"
+    nodesMsg n = "nodes : [" <> nodesTxt n <> "] already exist in current graphql schema"
+    nodesTxt nodes = T.intercalate ", " $ map G.unName nodes
+    builtin = ["__type", "__schema", "__typename", "Query", "Mutation", "query_root", "mutation_root"]
+    builtinTy = [ "__Directive"
```
Comment (0x777) [inline server/src-lib/Hasura/GraphQL/Schema.hs]: remove hardcoded `Query`, `Mutation` ... and filter query and mutation root types of both hasura and remote schema

File: server/src-lib/Hasura/GraphQL/Transport/HTTP.hs [hunk 8]
```diff
@@ -1,40 +1,170 @@
...
+                                          VQ.getQueryParts req
+
+  let topLevelNodes = getTopLevelNodes opDef
+      -- gather TypeLoc of topLevelNodes
+      typeLocs = gatherTypeLocs gCtx topLevelNodes
+
+  -- see if they are all the same
+  assertSameLocationNodes typeLocs
+
+  if null typeLocs
```
Comment (0x777) [inline server/src-lib/Hasura/GraphQL/Transport/HTTP.hs]: one single pattern match

File: server/src-lib/Hasura/GraphQL/Transport/HTTP.hs [hunk 9]
```diff
@@ -1,40 +1,170 @@
...
+    gCtxRoleMap = scGCtxMap sc
+
+assertSameLocationNodes :: (MonadError QErr m) => [VT.TypeLoc] -> m ()
+assertSameLocationNodes typeLocs =
+  -- see if they are all the same
+  unless (allEq typeLocs) $
+    throw400 NotSupported msg
+  where
+    msg = "cannot mix nodes from two different graphql servers"
+    allEq xs = case xs of
```
Comment (0x777) [inline server/src-lib/Hasura/GraphQL/Transport/HTTP.hs]: len(hashset(l)) == 1

File: server/src-lib/Hasura/GraphQL/Transport/WebSocket.hs [hunk 10]
```diff
@@ -54,7 +57,7 @@ type OperationMap
 data WSConnData
   = WSConnData
   -- the role and headers are set only on connection_init message
-  { _wscUser  :: !(IORef.IORef (Maybe (Either Text UserInfo)))
+  { _wscUser  :: !(IORef.IORef (Maybe (Either Text (UserInfo, [H.Header]))))
```
Comment (0x777) [inline server/src-lib/Hasura/GraphQL/Transport/WebSocket.hs]:
```haskell
data ConnSt 
  = CSNotInitialised
  | CSInitError
  | CSInitialised UserInfo [H.Header]
```

File: server/src-lib/Hasura/RQL/DDL/Schema/Table.hs [hunk 11]
```diff
@@ -137,25 +143,44 @@ trackExistingTableOrViewP1 (TrackTable vn) = do
...
   ti <- liftTx $ getTableInfo tn isSystemDefined
   addTableToCache ti
 
 trackExistingTableOrViewP2
-  :: (P2C m) => QualifiedTable -> Bool -> m RespBody
-trackExistingTableOrViewP2 vn isSystemDefined = do
+  :: (QErrM m, CacheRWM m, MonadTx m, MonadIO m, HasHttpManager m)
+  => QualifiedTable -> Bool -> Bool -> m RespBody
+trackExistingTableOrViewP2 vn isSystemDefined checkConflict = do
+  when checkConflict $ do
```
Comment (0x777) [inline server/src-lib/Hasura/RQL/DDL/Schema/Table.hs]: the checkconflict boolean is not needed

File: server/src-lib/Hasura/RQL/DDL/Schema/Table.hs [hunk 12]
```diff
@@ -137,25 +143,44 @@ trackExistingTableOrViewP1 (TrackTable vn) = do
...
+    sc <- askSchemaCache
+    let gCtxMap = scGCtxMap sc
+    forM_ (M.toList gCtxMap) $ \(_, gCtx) ->
```
Comment (0x777) [inline server/src-lib/Hasura/RQL/DDL/Schema/Table.hs]: only check against the merged remote (default) gctx

**Comment by shahidhk** [comment]
Tried adding GitHub API, 2 errors:

- GitHub needs a `User-Agent` header, we are not sending it.
- GitHub API contains interfaces, the error is a bit cryptic, better error message can help: `Error in $.types[11].interfaces[0]: expected Text, encountered Object`

**Comment by ecthiender** [comment]
@shahidhk Above two issues are fixed in the latest commit.

**Pull request merged**